docs/proposed/http-storage-node-protocol.rst

```diff
@@ -0,0 +1,506 @@
+For example::
+
+  {
+      "success": true,
+      "data": {
+          0: ["foo"],
+          5: ["bar"],
+          ...
+      }
+  }
+
+Reading
+~~~~~~~
+
+``GET /v1/mutable/:storage_index/shares``
+!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
+
+Retrieve a list indicating all shares available for the indicated storage index.
+For example::
+
+  [1, 5]
+
+``GET /v1/mutable/:storage_index?share=:s0&share=:sN&offset=o1&size=z0&offset=oN&size=zN``
+!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
+
+Read data from the indicated mutable shares.
+Just like ``GET /v1/mutable/:storage_index``.
+
+``POST /v1/mutable/:storage_index/:share_number/corrupt``
+!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
+
+Advise the server the data read from the indicated share was corrupt.
+Just like the immutable version.
+
+.. _RFC 7469: https://tools.ietf.org/html/rfc7469#section-2.4
+
+.. _RFC 7049: https://tools.ietf.org/html/rfc7049#section-4
+
+.. _CBOR: http://cbor.io/
+
+.. [#]
+   The security value of checking ``notValidBefore`` and ``notValidAfter`` is not entirely clear.
+   The arguments which apply to web-facing certificates do not seem to apply
+   (due to the decision for Tahoe-LAFS to operate independently of the web-oriented CA system).
+
+   There is an argument to make that complexity is reduced by allowing an existing TLS implementation which wants to make these checks make them
+   (compared to including additional code to either bypass them or disregard their results).
+   Reducing complexity, at least in general, is often good for security.
+
+   On the other hand, checking the validity time period forces certificate regeneration
+   (which comes with its own set of complexity).
+
+   A possible compromise is to recommend very certificates with validity periods of many years or decades.
+   "Recommend" may be read as "provide software supporting the generation of".
+
+   What about key theft?
+   If certificates are valid for years then a successful attacker can pretend to be a valid storage node for years.
+   However, short-validity-period certificates are no help in this case.
+   The attacker can generate new, valid certificates using the stolen keys.
+
+   Therefore, the only recourse to key theft
+   (really *identity theft*)
+   is to burn the identity and generate a new one.
+   Burning the identity is a non-trivial task.
+   It is worth solving but it is not solved here.
+
+.. [#]
+   More simply::
+
+    from hashlib import sha256
+    from cryptography.hazmat.primitives.serialization import (
+      Encoding,
+      PublicFormat,
+    )
+    from foolscap import base32
+
+    spki_bytes = cert.public_key().public_bytes(Encoding.DER, PublicFormat.SubjectPublicKeyInfo)
+    spki_sha256 = sha256(spki_bytes).digest()
+    spki_digest32 = base32.encode(spki_sha256)
+    assert spki_digest32 == tub_id
+
+   Note we use the Tahoe-LAFS-preferred base32 encoding rather than base64.
+
```

Comment:
Do we know why tahoe prefers base32 over more compact ones?

Reply:
I don't.  Maybe implementation ease.  Python stdlib doesn't offer URL-safe Base64, only regular Base64.  I think this is not a compelling reason at this point and there is a library on PyPI that does URL-safe Base64 so I'll switch from Base32 to that.